x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java

```diff
@@ -132,6 +165,89 @@ protected boolean skipExpression(Expression e) {
+
+        /**
+         * Local variant of the aggregation that returns the intermediate value.
+         */
+        @Override
+        protected void aggOutput(NamedExpression agg, AggregateFunction aggFunc, BlockFactory blockFactory, List<Block> blocks) {
+            List<Attribute> output = AbstractPhysicalOperationProviders.intermediateAttributes(List.of(agg), List.of());
+            for (Attribute o : output) {
+                DataType dataType = o.dataType();
+                // look for count(literal) with literal != null
+                var wrapper = BlockUtils.wrapperFor(blockFactory, PlannerUtils.toElementType(dataType), 1);
+                // seen output
+                if (dataType == DataTypes.BOOLEAN) {
+                    wrapper.accept(true);
+                } else {
+                    if (aggFunc instanceof Count count && (count.foldable() == false || count.fold() != null)) {
+                        wrapper.accept(0L);
+                    } else {
+                        // otherwise just put null
+                        wrapper.accept(null);
+                    }
+                }
+                blocks.add(wrapper.builder().build());
+            }
+        }
+    }
+
+    /**
+     * The vast majority of aggs ignore null entries - this rule adds a pushable filter, as it is cheap
+     * to execute, to filter this entries out to begin with.
+     * STATS x = min(a), y = sum(b)
+     * becomes
+     * | WHERE a IS NOT NULL OR b IS NOT NULL
+     * | STATS x = min(a), y = sum(b)
+     * <br>
+     * Unfortunately this optimization cannot be applied when grouping is necessary since it can filter out
+     * groups containing only null values
+     */
+    static class InferNonNullAggConstraint extends ParameterizedOptimizerRule<Aggregate, LocalLogicalOptimizerContext> {
+
+        @Override
+        protected LogicalPlan rule(Aggregate aggregate, LocalLogicalOptimizerContext context) {
+            // only look at aggregates with default grouping
+            if (aggregate.groupings().size() > 0) {
+                return aggregate;
+            }
+
+            SearchStats stats = context.searchStats();
+            LogicalPlan plan = aggregate;
+            var aggs = aggregate.aggregates();
+            Set<Expression> nonNullAggFields = Sets.newLinkedHashSetWithExpectedSize(aggs.size());
+            for (var agg : aggs) {
+                Expression expr = agg;
+                if (agg instanceof Alias as) {
+                    expr = as.child();
+                }
+                if (expr instanceof AggregateFunction af) {
+                    Expression field = af.field();
+                    // ignore literals (e.g. COUNT(1))
+                    // make sure the field exists at the source and is indexed (not runtime)
+                    if (field.foldable() == false && field instanceof FieldAttribute fa && stats.isIndexed(fa.name())) {
```

Comment:
The filter is pushed down only if all the fields are indexed (this excludes scripted fields) - thanks to @dnhatn for pointing this out.
Raised #103699 in relation to this.